File: backend/app/services/layout_validation_service.py
"""
Comprehensive validation for manually adjusted pallet layouts.

All geometry uses mm.  Boxes are axis-aligned rectangular prisms.
"""
import re
from typing import List, Dict, Optional
from app.core.schemas import (
    LayoutPatchRequest, AdjustmentValidation, BoxResult,
    ManualAdjustmentSettings, StabilityIssue,
)

SUPPORT_THRESHOLD = 0.70   # 70 % of base area must be supported


def _issue_type(message: str) -> str:
    msg = message.lower()
    if "floating" in msg or "support" in msg:
        return "floating" if "no support" in msg or "floating" in msg else "unstable"
    if "overlap" in msg:
        return "overlap"
    if "threshold" in msg or "edge" in msg or "outside" in msg:
        return "boundary"
    if "height" in msg:
        return "height"
    if "no_load_on_top" in msg:
        return "no_load_on_top"
    if "stand_upright_only" in msg:
        return "stand_upright_only"
    if "duplicate" in msg:
        return "duplicate"
    if "missing" in msg:
        return "missing"
    return "validation"


def _box_ids_from_message(message: str) -> List[str]:
    ids = re.findall(r"[Bb]ox(?:es)? '?([A-Za-z0-9_-]+)'?", message)
    ids.extend(re.findall(r"and '([A-Za-z0-9_-]+)'", message))
    ids.extend(re.findall(r"of '([A-Za-z0-9_-]+)'", message))
    ids.extend(re.findall(r"Duplicate box_id '([A-Za-z0-9_-]+)'", message))
    return list(dict.fromkeys(ids))


def _issues_from_messages(
    errors: List[str],
    warnings: List[str],
    box_meta: Dict[str, tuple[int, Optional[str]]],
) -> List[StabilityIssue]:
    issues: List[StabilityIssue] = []
    for severity, messages in (("error", errors), ("warning", warnings)):
        for message in messages:
            ids = _box_ids_from_message(message)
            if not ids:
                issues.append(
                    StabilityIssue(
                        type=_issue_type(message),
                        severity=severity,
                        message=message,
                    )
                )
                continue
            for bid in ids:
                pallet_no, sku = box_meta.get(bid, (None, None))
                issues.append(
                    StabilityIssue(
                        type=_issue_type(message),
                        severity=severity,
                        box_id=bid,
                        pallet_no=pallet_no,
                        sku=sku,
                        message=message,
                    )
                )
    return issues


# ─────────────────────────────────────────────────────────────────────────────
# Geometry helpers
# ─────────────────────────────────────────────────────────────────────────────

def _overlaps(ax, ay, az, al, aw, ah, bx, by, bz, bl, bw, bh) -> bool:
    """Return True if two boxes have overlapping (non-touching) volume."""
    EPS = 0.5  # 0.5 mm tolerance — touching faces are OK
    return (
        ax < bx + bl - EPS and ax + al > bx + EPS
        and ay < by + bw - EPS and ay + aw > by + EPS
        and az < bz + bh - EPS and az + ah > bz + EPS
    )


def _xy_overlap_area(ax, ay, al, aw, bx, by, bl, bw) -> float:
    """Return the overlapping area in the XY plane between two box footprints."""
    ox = min(ax + al, bx + bl) - max(ax, bx)
    oy = min(ay + aw, by + bw) - max(ay, by)
    if ox <= 0 or oy <= 0:
        return 0.0
    return ox * oy


# ─────────────────────────────────────────────────────────────────────────────
# Individual check functions
# ─────────────────────────────────────────────────────────────────────────────

def check_boundary_with_threshold(
    boxes: List[dict],
    pallet: dict,
    settings: ManualAdjustmentSettings,
    errors: List[str],
) -> None:
    tl = settings.edge_threshold_length_mm
    tw = settings.edge_threshold_width_mm
    L = pallet["length_mm"]
    W = pallet["width_mm"]

    for b in boxes:
        bid = b["box_id"]
        x, y = b["x_mm"], b["y_mm"]
        l, w = b["length_mm"], b["width_mm"]

        if x < tl - 0.5:
            errors.append(f"Box '{bid}': x={x:.1f} violates length edge threshold ({tl} mm).")
        if y < tw - 0.5:
            errors.append(f"Box '{bid}': y={y:.1f} violates width edge threshold ({tw} mm).")
        if x + l > L - tl + 0.5:
            errors.append(f"Box '{bid}': right edge {x+l:.1f} violates length edge threshold (max {L-tl:.1f}).")
        if y + w > W - tw + 0.5:
            errors.append(f"Box '{bid}': front edge {y+w:.1f} violates width edge threshold (max {W-tw:.1f}).")


def check_max_height(boxes: List[dict], pallet: dict, errors: List[str]) -> None:
    H = pallet["max_height_mm"]
    for b in boxes:
        top = b["z_mm"] + b["height_mm"]
        if top > H + 0.5:
            errors.append(f"Box '{b['box_id']}': top {top:.1f} mm exceeds pallet max height {H} mm.")


def check_max_weight(
    boxes: List[dict],
    pallet_no: int,
    pallet: dict,
    errors: List[str],
) -> None:
    total = sum(b.get("weight_kg", 0) for b in boxes)
    if total > pallet["max_weight_kg"] + 0.001:
        errors.append(
            f"Pallet {pallet_no}: total weight {total:.2f} kg exceeds max {pallet['max_weight_kg']} kg."
        )


def check_overlap(boxes: List[dict], pallet_no: int, errors: List[str]) -> None:
    for i in range(len(boxes)):
        for j in range(i + 1, len(boxes)):
            a, b = boxes[i], boxes[j]
            if _overlaps(
                a["x_mm"], a["y_mm"], a["z_mm"], a["length_mm"], a["width_mm"], a["height_mm"],
                b["x_mm"], b["y_mm"], b["z_mm"], b["length_mm"], b["width_mm"], b["height_mm"],
            ):
                errors.append(
                    f"Pallet {pallet_no}: boxes '{a['box_id']}' and '{b['box_id']}' overlap."
                )


def check_floating(
    boxes: List[dict],
    pallet_no: int,
    errors: List[str],
    warnings: List[str],
    strict_stability: bool,
) -> None:
    """Box at z=0 is on the floor.  z>0 requires ≥70% XY support from boxes directly below."""
    for box in boxes:
        if box["z_mm"] < 0.5:
            continue  # floor

        base_area = box["length_mm"] * box["width_mm"]
        if base_area == 0:
            continue

        supported = 0.0
        for other in boxes:
            if other["box_id"] == box["box_id"]:
                continue
            if abs((other["z_mm"] + other["height_mm"]) - box["z_mm"]) < 1.0:
                supported += _xy_overlap_area(
                    box["x_mm"], box["y_mm"], box["length_mm"], box["width_mm"],
                    other["x_mm"], other["y_mm"], other["length_mm"], other["width_mm"],
                )

        frac = min(supported / base_area, 1.0)
        if frac == 0.0:
            message = (
                f"Pallet {pallet_no}: box '{box['box_id']}' is floating "
                f"(z={box['z_mm']:.1f}, no support below)."
            )
            (errors if strict_stability else warnings).append(message)
        elif frac < SUPPORT_THRESHOLD:
            message = (
                f"Pallet {pallet_no}: box '{box['box_id']}' has only "
                f"{frac*100:.0f}% support (minimum {SUPPORT_THRESHOLD*100:.0f}%)."
            )
            (errors if strict_stability else warnings).append(message)


def check_no_load_on_top(
    boxes: List[dict],
    orig_map: Dict[str, BoxResult],
    pallet_no: int,
    errors: List[str],
    warnings: List[str],
    strict_stability: bool,
) -> None:
    """If box has no_load_on_top, nothing may rest on its top face."""
    for box in boxes:
        orig = orig_map.get(box["box_id"])
        no_load = (orig.no_load_on_top if orig else False) or box.get("no_load_on_top", False)
        if not no_load:
            continue
        top_z = box["z_mm"] + box["height_mm"]
        for other in boxes:
            if other["box_id"] == box["box_id"]:
                continue
            if abs(other["z_mm"] - top_z) < 1.0:
                overlap = _xy_overlap_area(
                    box["x_mm"], box["y_mm"], box["length_mm"], box["width_mm"],
                    other["x_mm"], other["y_mm"], other["length_mm"], other["width_mm"],
                )
                if overlap > 1.0:
                    message = (
                        f"Pallet {pallet_no}: box '{other['box_id']}' is on top of "
                        f"'{box['box_id']}' which has no_load_on_top constraint."
                    )
                    (errors if strict_stability else warnings).append(message)


def check_stand_upright_only(
    boxes: List[dict],
    orig_map: Dict[str, BoxResult],
    pallet_no: int,
    errors: List[str],
) -> None:
    """
    If box has stand_upright_only, its height_mm must equal original_height_mm.
    This prevents rotations that would flip the upright axis.
    """
    for box in boxes:
        orig = orig_map.get(box["box_id"])
        upright = (orig.stand_upright_only if orig else False) or box.get("stand_upright_only", False)
        if not upright:
            continue
        orig_h = orig.original_height_mm if orig and orig.original_height_mm else (orig.height_mm if orig else None)
        if orig_h is None:
            continue
        if abs(box["height_mm"] - orig_h) > 0.5:
            errors.append(
                f"Pallet {pallet_no}: box '{box['box_id']}' has stand_upright_only constraint "
                f"but height changed from {orig_h:.1f} to {box['height_mm']:.1f} mm."
            )


def check_duplicate_box_ids(
    all_boxes: List[dict],
    errors: List[str],
) -> None:
    seen: set = set()
    for b in all_boxes:
        bid = b["box_id"]
        if bid in seen:
            errors.append(f"Duplicate box_id '{bid}' found in adjusted layout.")
        seen.add(bid)


def check_missing_and_extra_boxes(
    submitted_ids: set,
    original_ids: set,
    warnings: List[str],
    errors: List[str],
) -> None:
    for bid in sorted(original_ids - submitted_ids):
        warnings.append(f"Box '{bid}' from generated layout is missing in adjusted layout.")
    for bid in sorted(submitted_ids - original_ids):
        errors.append(f"Box '{bid}' is not in the original generated layout (unknown box).")


# ─────────────────────────────────────────────────────────────────────────────
# Top-level validator
# ─────────────────────────────────────────────────────────────────────────────

def validate_adjusted_layout(
    layout: LayoutPatchRequest,
    pallet_config: dict,
    original_box_ids: set,
    orig_box_map: Dict[str, BoxResult],
) -> AdjustmentValidation:
    """
    Full validation of a manually adjusted layout.

    pallet_config  keys: length_mm, width_mm, max_height_mm, max_weight_kg
    original_box_ids: set of box_ids from the generated result
    orig_box_map: box_id → original BoxResult (for constraint lookup)
    """
    errors: List[str] = []
    warnings: List[str] = []

    all_boxes: List[dict] = []
    submitted_ids: set = set()
    box_meta: Dict[str, tuple[int, Optional[str]]] = {}

    for pallet in layout.pallets:
        boxes = [b.model_dump() for b in pallet.boxes]
        all_boxes.extend(boxes)
        for b in boxes:
            orig = orig_box_map.get(b["box_id"])
            box_meta[b["box_id"]] = (pallet.pallet_no, orig.sku if orig else None)

        check_boundary_with_threshold(boxes, pallet_config, layout.settings, errors)
        check_max_height(boxes, pallet_config, errors)
        check_max_weight(boxes, pallet.pallet_no, pallet_config, errors)
        check_overlap(boxes, pallet.pallet_no, errors)
        check_floating(
            boxes,
            pallet.pallet_no,
            errors,
            warnings,
            layout.settings.do_not_allow_stability_issues,
        )
        check_no_load_on_top(
            boxes,
            orig_box_map,
            pallet.pallet_no,
            errors,
            warnings,
            layout.settings.do_not_allow_stability_issues,
        )
        check_stand_upright_only(boxes, orig_box_map, pallet.pallet_no, errors)

        for b in boxes:
            submitted_ids.add(b["box_id"])

    check_duplicate_box_ids(all_boxes, errors)
    check_missing_and_extra_boxes(submitted_ids, original_box_ids, warnings, errors)
    issues = _issues_from_messages(errors, warnings, box_meta)
    invalid_box_ids = sorted({
        issue.box_id
        for issue in issues
        if issue.box_id and issue.severity == "error"
    })

    return AdjustmentValidation(
        is_valid=len(errors) == 0,
        invalidBoxIds=invalid_box_ids,
        issues=issues,
        errors=errors,
        warnings=warnings,
    )

File: palletisation-java/src/main/java/com/psap/palletisation/service/LayoutValidationService.java
package com.psap.palletisation.service;

import com.psap.palletisation.dto.request.BoxPatch;
import com.psap.palletisation.dto.request.LayoutPatchRequest;
import com.psap.palletisation.dto.request.ManualAdjustmentSettings;
import com.psap.palletisation.dto.request.PalletPatch;
import com.psap.palletisation.dto.response.AdjustmentValidation;
import com.psap.palletisation.dto.response.BoxResult;
import com.psap.palletisation.dto.response.StabilityIssue;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Comprehensive validation for manually adjusted layouts.
 * Port of layout_validation_service.py.
 */
@Service
public class LayoutValidationService {

    private static final double SUPPORT_THRESHOLD = 0.70;

    public AdjustmentValidation validateAdjustedLayout(
            LayoutPatchRequest layout,
            Map<String, Object> palletConfig,
            Set<String> originalBoxIds,
            Map<String, BoxResult> origBoxMap) {

        List<String> errors = new ArrayList<>();
        List<String> warnings = new ArrayList<>();
        List<Map<String, Object>> allBoxes = new ArrayList<>();
        Set<String> submittedIds = new LinkedHashSet<>();
        Map<String, int[]> boxMeta = new LinkedHashMap<>(); // boxId → [palletNo, sku_placeholder]

        double L = toDouble(palletConfig.get("length_mm"));
        double W = toDouble(palletConfig.get("width_mm"));
        double H = toDouble(palletConfig.get("max_height_mm"));
        double maxW = toDouble(palletConfig.get("max_weight_kg"));

        ManualAdjustmentSettings settings = layout.getSettings();

        for (PalletPatch palletPatch : layout.getPallets()) {
            List<Map<String, Object>> boxes = toBoxMaps(palletPatch.getBoxes());
            allBoxes.addAll(boxes);
            for (Map<String, Object> b : boxes) {
                String bid = (String) b.get("box_id");
                boxMeta.put(bid, new int[]{palletPatch.getPalletNo()});
            }

            checkBoundaryWithThreshold(boxes, palletConfig, settings, errors);
            checkMaxHeight(boxes, palletConfig, errors);
            checkMaxWeight(boxes, palletPatch.getPalletNo(), palletConfig, errors);
            checkOverlap(boxes, palletPatch.getPalletNo(), errors);
            checkFloating(boxes, palletPatch.getPalletNo(), errors, warnings, settings.isDoNotAllowStabilityIssues());
            checkNoLoadOnTop(boxes, origBoxMap, palletPatch.getPalletNo(), errors, warnings,
                    settings.isEnforceNoLoadOnTop(), settings.isDoNotAllowStabilityIssues());
            checkStandUprightOnly(boxes, origBoxMap, palletPatch.getPalletNo(), errors);

            for (Map<String, Object> b : boxes) {
                submittedIds.add((String) b.get("box_id"));
            }
        }

        checkDuplicateBoxIds(allBoxes, errors);
        checkMissingAndExtraBoxes(submittedIds, originalBoxIds, warnings, errors);

        // Build issues from error/warning messages
        Map<String, Object[]> boxMetaForIssues = new LinkedHashMap<>();
        for (Map.Entry<String, int[]> e : boxMeta.entrySet()) {
            BoxResult orig = origBoxMap.get(e.getKey());
            boxMetaForIssues.put(e.getKey(), new Object[]{e.getValue()[0], orig != null ? orig.getSku() : null});
        }
        List<StabilityIssue> issues = issuesFromMessages(errors, warnings, boxMetaForIssues);
        List<String> invalidBoxIds = issues.stream()
                .filter(i -> i.getBoxId() != null && "error".equals(i.getSeverity()))
                .map(StabilityIssue::getBoxId)
                .distinct()
                .sorted()
                .toList();

        AdjustmentValidation result = new AdjustmentValidation();
        result.setIsValid(errors.isEmpty());
        result.setInvalidBoxIds(new ArrayList<>(invalidBoxIds));
        result.setIssues(issues);
        result.setErrors(errors);
        result.setWarnings(warnings);
        return result;
    }

    // ── Check functions ───────────────────────────────────────────────────────

    private void checkBoundaryWithThreshold(List<Map<String, Object>> boxes,
                                              Map<String, Object> palletConfig,
                                              ManualAdjustmentSettings settings,
                                              List<String> errors) {
        double tl = settings.getEdgeThresholdLengthMm();
        double tw = settings.getEdgeThresholdWidthMm();
        double L = toDouble(palletConfig.get("length_mm"));
        double W = toDouble(palletConfig.get("width_mm"));

        for (Map<String, Object> b : boxes) {
            String bid = (String) b.get("box_id");
            double x = toDouble(b.get("x_mm")), y = toDouble(b.get("y_mm"));
            double l = toDouble(b.get("length_mm")), w = toDouble(b.get("width_mm"));
            if (x < tl - 0.5) errors.add(String.format("Box '%s': x=%.1f violates length edge threshold (%.1f mm).", bid, x, tl));
            if (y < tw - 0.5) errors.add(String.format("Box '%s': y=%.1f violates width edge threshold (%.1f mm).", bid, y, tw));
            if (x + l > L - tl + 0.5) errors.add(String.format("Box '%s': right edge %.1f violates length edge threshold (max %.1f).", bid, x + l, L - tl));
            if (y + w > W - tw + 0.5) errors.add(String.format("Box '%s': front edge %.1f violates width edge threshold (max %.1f).", bid, y + w, W - tw));
        }
    }

    private void checkMaxHeight(List<Map<String, Object>> boxes, Map<String, Object> palletConfig, List<String> errors) {
        double H = toDouble(palletConfig.get("max_height_mm"));
        for (Map<String, Object> b : boxes) {
            double top = toDouble(b.get("z_mm")) + toDouble(b.get("height_mm"));
            if (top > H + 0.5)
                errors.add(String.format("Box '%s': top %.1f mm exceeds pallet max height %.1f mm.", b.get("box_id"), top, H));
        }
    }

    private void checkMaxWeight(List<Map<String, Object>> boxes, int palletNo,
                                  Map<String, Object> palletConfig, List<String> errors) {
        double total = boxes.stream().mapToDouble(b -> toDouble(b.getOrDefault("weight_kg", 0))).sum();
        double maxW = toDouble(palletConfig.get("max_weight_kg"));
        if (total > maxW + 0.001)
            errors.add(String.format("Pallet %d: total weight %.2f kg exceeds max %.1f kg.", palletNo, total, maxW));
    }

    private void checkOverlap(List<Map<String, Object>> boxes, int palletNo, List<String> errors) {
        for (int i = 0; i < boxes.size(); i++) {
            for (int j = i + 1; j < boxes.size(); j++) {
                Map<String, Object> a = boxes.get(i), b = boxes.get(j);
                if (overlaps(a, b))
                    errors.add(String.format("Pallet %d: boxes '%s' and '%s' overlap.", palletNo, a.get("box_id"), b.get("box_id")));
            }
        }
    }

    private void checkFloating(List<Map<String, Object>> boxes, int palletNo,
                                 List<String> errors, List<String> warnings, boolean strict) {
        for (Map<String, Object> box : boxes) {
            if (toDouble(box.get("z_mm")) < 0.5) continue;
            double baseArea = toDouble(box.get("length_mm")) * toDouble(box.get("width_mm"));
            if (baseArea == 0) continue;
            double supported = 0;
            for (Map<String, Object> other : boxes) {
                if (other.get("box_id").equals(box.get("box_id"))) continue;
                if (Math.abs((toDouble(other.get("z_mm")) + toDouble(other.get("height_mm"))) - toDouble(box.get("z_mm"))) < 1.0) {
                    supported += xyOverlapArea(
                            toDouble(box.get("x_mm")), toDouble(box.get("y_mm")),
                            toDouble(box.get("length_mm")), toDouble(box.get("width_mm")),
                            toDouble(other.get("x_mm")), toDouble(other.get("y_mm")),
                            toDouble(other.get("length_mm")), toDouble(other.get("width_mm")));
                }
            }
            double frac = Math.min(supported / baseArea, 1.0);
            if (frac == 0.0) {
                String msg = String.format("Pallet %d: box '%s' is floating (z=%.1f, no support below).",
                        palletNo, box.get("box_id"), toDouble(box.get("z_mm")));
                (strict ? errors : warnings).add(msg);
            } else if (frac < SUPPORT_THRESHOLD) {
                String msg = String.format("Pallet %d: box '%s' has only %.0f%% support (minimum %.0f%%).",
                        palletNo, box.get("box_id"), frac * 100, SUPPORT_THRESHOLD * 100);
                (strict ? errors : warnings).add(msg);
            }
        }
    }

    private void checkNoLoadOnTop(List<Map<String, Object>> boxes, Map<String, BoxResult> origMap,
                                    int palletNo, List<String> errors, List<String> warnings,
                                    boolean enforceNoLoadOnTop, boolean strict) {
        if (!enforceNoLoadOnTop) return;
        for (Map<String, Object> box : boxes) {
            BoxResult orig = origMap.get(box.get("box_id"));
            boolean noLoad = (orig != null && orig.isNoLoadOnTop()) ||
                    Boolean.TRUE.equals(box.get("no_load_on_top"));
            if (!noLoad) continue;
            double topZ = toDouble(box.get("z_mm")) + toDouble(box.get("height_mm"));
            for (Map<String, Object> other : boxes) {
                if (other.get("box_id").equals(box.get("box_id"))) continue;
                if (Math.abs(toDouble(other.get("z_mm")) - topZ) < 1.0) {
                    double overlap = xyOverlapArea(
                            toDouble(box.get("x_mm")), toDouble(box.get("y_mm")),
                            toDouble(box.get("length_mm")), toDouble(box.get("width_mm")),
                            toDouble(other.get("x_mm")), toDouble(other.get("y_mm")),
                            toDouble(other.get("length_mm")), toDouble(other.get("width_mm")));
                    if (overlap > 1.0) {
                        String msg = String.format("Pallet %d: box '%s' is on top of '%s' which has no_load_on_top constraint.",
                                palletNo, other.get("box_id"), box.get("box_id"));
                        (strict ? errors : warnings).add(msg);
                    }
                }
            }
        }
    }

    private void checkStandUprightOnly(List<Map<String, Object>> boxes, Map<String, BoxResult> origMap,
                                         int palletNo, List<String> errors) {
        for (Map<String, Object> box : boxes) {
            BoxResult orig = origMap.get(box.get("box_id"));
            boolean upright = (orig != null && orig.isStandUprightOnly()) ||
                    Boolean.TRUE.equals(box.get("stand_upright_only"));
            if (!upright) continue;
            Double origH = orig != null ? (orig.getOriginalHeightMm() != null ? orig.getOriginalHeightMm() : orig.getHeightMm()) : null;
            if (origH == null) continue;
            if (Math.abs(toDouble(box.get("height_mm")) - origH) > 0.5) {
                errors.add(String.format("Pallet %d: box '%s' has stand_upright_only constraint but height changed from %.1f to %.1f mm.",
                        palletNo, box.get("box_id"), origH, toDouble(box.get("height_mm"))));
            }
        }
    }

    private void checkDuplicateBoxIds(List<Map<String, Object>> allBoxes, List<String> errors) {
        Set<String> seen = new LinkedHashSet<>();
        for (Map<String, Object> b : allBoxes) {
            String bid = (String) b.get("box_id");
            if (!seen.add(bid))
                errors.add(String.format("Duplicate box_id '%s' found in adjusted layout.", bid));
        }
    }

    private void checkMissingAndExtraBoxes(Set<String> submitted, Set<String> original,
                                             List<String> warnings, List<String> errors) {
        Set<String> missing = new TreeSet<>(original);
        missing.removeAll(submitted);
        Set<String> extra = new TreeSet<>(submitted);
        extra.removeAll(original);
        for (String bid : missing)
            warnings.add(String.format("Box '%s' from generated layout is missing in adjusted layout.", bid));
        for (String bid : extra)
            errors.add(String.format("Box '%s' is not in the original generated layout (unknown box).", bid));
    }

    // ── Geometry helpers ──────────────────────────────────────────────────────

    private boolean overlaps(Map<String, Object> a, Map<String, Object> b) {
        double EPS = 0.5;
        double ax = toDouble(a.get("x_mm")), ay = toDouble(a.get("y_mm")), az = toDouble(a.get("z_mm"));
        double al = toDouble(a.get("length_mm")), aw = toDouble(a.get("width_mm")), ah = toDouble(a.get("height_mm"));
        double bx = toDouble(b.get("x_mm")), by = toDouble(b.get("y_mm")), bz = toDouble(b.get("z_mm"));
        double bl = toDouble(b.get("length_mm")), bw = toDouble(b.get("width_mm")), bh = toDouble(b.get("height_mm"));
        return ax < bx + bl - EPS && ax + al > bx + EPS
                && ay < by + bw - EPS && ay + aw > by + EPS
                && az < bz + bh - EPS && az + ah > bz + EPS;
    }

    private double xyOverlapArea(double ax, double ay, double al, double aw,
                                   double bx, double by, double bl, double bw) {
        double ox = Math.min(ax + al, bx + bl) - Math.max(ax, bx);
        double oy = Math.min(ay + aw, by + bw) - Math.max(ay, by);
        return (ox > 0 && oy > 0) ? ox * oy : 0.0;
    }

    // ── Issue builder ─────────────────────────────────────────────────────────

    private List<StabilityIssue> issuesFromMessages(List<String> errors, List<String> warnings,
                                                      Map<String, Object[]> boxMeta) {
        List<StabilityIssue> issues = new ArrayList<>();
        for (String[] entry : new String[][]{}) {} // placeholder
        for (int pass = 0; pass < 2; pass++) {
            String severity = pass == 0 ? "error" : "warning";
            List<String> messages = pass == 0 ? errors : warnings;
            for (String message : messages) {
                List<String> ids = extractBoxIds(message);
                if (ids.isEmpty()) {
                    StabilityIssue issue = new StabilityIssue();
                    issue.setType(issueType(message));
                    issue.setSeverity(severity);
                    issue.setMessage(message);
                    issues.add(issue);
                } else {
                    for (String bid : ids) {
                        Object[] meta = boxMeta.get(bid);
                        StabilityIssue issue = new StabilityIssue();
                        issue.setType(issueType(message));
                        issue.setSeverity(severity);
                        issue.setBoxId(bid);
                        if (meta != null) {
                            issue.setPalletNo((Integer) meta[0]);
                            issue.setSku((String) meta[1]);
                        }
                        issue.setMessage(message);
                        issues.add(issue);
                    }
                }
            }
        }
        return issues;
    }

    private static final Pattern BOX_ID_PATTERN = Pattern.compile("'([A-Za-z0-9_-]+)'");

    private List<String> extractBoxIds(String message) {
        List<String> ids = new ArrayList<>();
        Matcher m = BOX_ID_PATTERN.matcher(message);
        while (m.find()) {
            ids.add(m.group(1));
        }
        return ids.stream().distinct().toList();
    }

    private String issueType(String message) {
        String msg = message.toLowerCase();
        if (msg.contains("floating") || msg.contains("no support")) return "floating";
        if (msg.contains("support")) return "unstable";
        if (msg.contains("overlap")) return "overlap";
        if (msg.contains("threshold") || msg.contains("edge") || msg.contains("outside") || msg.contains("exceeds")) return "boundary";
        if (msg.contains("height")) return "height";
        if (msg.contains("no_load_on_top")) return "no_load_on_top";
        if (msg.contains("stand_upright_only")) return "stand_upright_only";
        if (msg.contains("duplicate")) return "duplicate";
        if (msg.contains("missing")) return "missing";
        return "validation";
    }

    // ── Conversion helpers ────────────────────────────────────────────────────

    private double toDouble(Object val) {
        if (val instanceof Number) return ((Number) val).doubleValue();
        if (val == null) return 0.0;
        try { return Double.parseDouble(val.toString()); } catch (NumberFormatException e) { return 0.0; }
    }

    private List<Map<String, Object>> toBoxMaps(List<BoxPatch> patches) {
        List<Map<String, Object>> result = new ArrayList<>();
        for (BoxPatch bp : patches) {
            Map<String, Object> m = new LinkedHashMap<>();
            m.put("box_id", bp.getBoxId());
            m.put("x_mm", bp.getXMm());
            m.put("y_mm", bp.getYMm());
            m.put("z_mm", bp.getZMm());
            m.put("length_mm", bp.getLengthMm());
            m.put("width_mm", bp.getWidthMm());
            m.put("height_mm", bp.getHeightMm());
            m.put("rotation", bp.getRotation());
            m.put("layer", bp.getLayer());
            m.put("stand_upright_only", bp.isStandUprightOnly());
            m.put("no_load_on_top", bp.isNoLoadOnTop());
            m.put("weight_kg", 0.0); // weight not in patch
            result.add(m);
        }
        return result;
    }
}
